src/backend/distributed/utils/resource_lock.c

```diff
@@ -316,3 +318,69 @@ LockRelationShardResources(List *relationShardList, LOCKMODE lockMode)
 		}
 	}
 }
+
+
+/*
+ * LockParentShardResourceIfPartition acquires a shard resource lock on colocated
+ * shard of parent table of given shard's relation if given shard belongs to a
+ * partition.
+ */
+void
+LockParentShardResourceIfPartition(uint64 shardId, LOCKMODE lockMode)
+{
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	Oid relationId = shardInterval->relationId;
+
+	if (PartitionTable(relationId))
+	{
+		int shardIndex = ShardIndex(shardInterval);
+		Oid parentRelationId = PartitionParentOid(relationId);
+		uint64 parentShardId = ColocatedShardIdInRelation(parentRelationId, shardIndex);
+
+		LockShardResource(parentShardId, lockMode);
+	}
+}
+
+
+/*
+ * LockPartitionsInQuery traverses the query and extracts relations in it. Then
+ * it acquires locks on partitions of each partitioned table. It does nothing
+ * for non-partitioned tables.
+ */
+void
+LockPartitionsInRelationList(List *relationIdList, LOCKMODE lockmode)
+{
+	ListCell *relationIdCell = NULL;
+
+	foreach(relationIdCell, relationIdList)
+	{
+		Oid relationId = lfirst_oid(relationIdCell);
+		if (PartitionedTable(relationId))
+		{
+			LockPartitionRelations(relationId, lockmode);
+		}
+	}
+}
+
+
+/*
+ * LockPartitionRelations acquires relation lock on all partitions of given
```

Comment:
I added the comment but do we really need Assert? In the PartitionList function we call, we already error out if given table is not a partitioned table.